include/deal.II/matrix_free/dof_info.templates.h

```diff
@@ -1664,188 +659,166 @@ namespace internal
+     DynamicSparsityPattern          &connectivity) const
     {
-      AssertDimension (row_starts.size()-1, size_info.n_active_cells);
-      const unsigned int n_rows =
+      unsigned int n_rows =
         (vector_partitioner->local_range().second-
          vector_partitioner->local_range().first)
         + vector_partitioner->ghost_indices().n_elements();
-      const unsigned int n_blocks = (do_blocking == true) ?
-                                    task_info.n_blocks : size_info.n_active_cells;
+
+      // Avoid square sparsity patterns that allocate the diagonal entry
+      if (n_rows == task_info.n_active_cells)
+        ++n_rows;
 
       // first determine row lengths
       std::vector<unsigned int> row_lengths(n_rows);
-      unsigned int cell_start = 0, mcell_start = 0;
-      std::vector<unsigned int> scratch;
-      for (unsigned int block = 0; block < n_blocks; ++block)
-        {
-          // if we have the blocking variant (used in the coloring scheme), we
-          // want to build a graph with the blocks with interaction with
-          // remote MPI processes up front. in the non-blocking variant, we do
-          // not do this here. TODO: unify this approach!!!
-          if (do_blocking == true)
-            {
-              scratch.clear();
-              for (unsigned int mcell=mcell_start; mcell<
-                   std::min(mcell_start+task_info.block_size,
-                            size_info.n_macro_cells);
-                   ++mcell)
-                {
-                  unsigned int n_comp = (irregular_cells[mcell]>0)
-                                        ?irregular_cells[mcell]:size_info.vectorization_length;
-                  for (unsigned int cell = cell_start; cell < cell_start+n_comp;
-                       ++cell)
-                    scratch.insert(scratch.end(),
-                                   begin_indices(renumbering[cell]),
-                                   end_indices(renumbering[cell]));
-                  cell_start += n_comp;
-                }
-              std::sort(scratch.begin(), scratch.end());
-              const unsigned int n_unique =
-                std::unique(scratch.begin(), scratch.end())-scratch.begin();
-              for (unsigned int i=0; i<n_unique; ++i)
-                row_lengths[scratch[i]]++;
-              mcell_start += task_info.block_size;
-            }
-          else
-            {
-              scratch.clear();
-              scratch.insert(scratch.end(),
-                             begin_indices(block), end_indices(block));
-              std::sort(scratch.begin(), scratch.end());
-              const unsigned int n_unique =
-                std::unique(scratch.begin(), scratch.end())-scratch.begin();
-              for (unsigned int i=0; i<n_unique; ++i)
-                row_lengths[scratch[i]]++;
-            }
-        }
-
-      // disregard dofs that only sit on one cell
+      std::vector<Threads::Mutex> mutexes(n_rows/256+1);
```

Comment:
the split into 256 is used in several places, maybe declare a static constexpr constant and use it throughout?